pkg/rbac/decorator.go

```diff
@@ -0,0 +1,145 @@
+		return nil, err
+	}
+	res.GasPayment += cost
+	return res, nil
+}
+
+// Deliver enforces roles added to the context before calling down the stack
+func (d AuthNDecorator) Deliver(ctx weave.Context, store weave.KVStore, tx weave.Tx, next weave.Deliverer) (*weave.DeliverResult, error) {
+	newCtx, _, err := d.authRoles(ctx, store)
+	if err != nil {
+		return nil, err
+	}
+
+	return next.Deliver(newCtx, store, tx)
+}
+
+func (d AuthNDecorator) authRoles(ctx weave.Context, db weave.KVStore) (weave.Context, int64, error) {
+	var costs int64
+	roleIdsProcessed := make(map[string]Role)
+	for _, c := range d.authN.GetConditions(ctx) {
+		roleIDs, err := d.roleBinBucket.FindRoleIDsByAddress(db, c.Address())
+		if err != nil {
+			return ctx, 0, err
+		}
+		for _, roleID := range roleIDs {
+			err := d.loadRoles(db, roleID, roleIdsProcessed)
+			if err != nil {
+				return ctx, 0, err
+			}
+			costs += roleParticipantGasCost
+		}
+	}
+	if len(roleIdsProcessed) != 0 {
+		ctx = withRBAC(ctx, roleIdsProcessed)
+	}
+	return ctx, costs, nil
+}
+
+func (d AuthNDecorator) loadRoles(db weave.KVStore, roleID []byte, roleIdsProcessed map[string]Role) error {
+	if _, ok := roleIdsProcessed[string(roleID)]; ok {
+		return nil
+	}
+	var r Role
+	if err := d.roleBucket.One(db, roleID, &r); err != nil {
+		return err
+	}
+	roleIdsProcessed[string(roleID)] = r
+
+	for _, id := range r.RoleIds {
+		err := d.loadRoles(db, id, roleIdsProcessed)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type Authorizator interface {
+	HasPermission(ctx weave.Context, p Permission) bool
+}
+
+// AuthZDecorator handles authorization
+type AuthZDecorator struct {
+	authZ            Authorizator
+	permissionPrefix string
+}
+
+func NewAuthZDecorator(auth Authorizator, permissionPrefix string) AuthZDecorator {
+	return AuthZDecorator{
+		authZ:            auth,
+		permissionPrefix: permissionPrefix,
+	}
+}
+
+// Check enforces permissions for the message stored in the context before calling down the stack
+func (d AuthZDecorator) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx, next weave.Checker) (*weave.CheckResult, error) {
+	msg, err := tx.GetMsg()
+	if err != nil {
+		return nil, err
+	}
+	perm := d.resolvePermission(msg)
+	if !d.authZ.HasPermission(ctx, perm) {
+		return nil, errors.Wrap(errors.ErrUnauthorized, "insufficient permissions")
+	}
+	return next.Check(ctx, store, tx)
+}
+
+// Deliver enforces permissions for the message stored in the context before calling down the stack
+func (d AuthZDecorator) Deliver(ctx weave.Context, store weave.KVStore, tx weave.Tx, next weave.Deliverer) (*weave.DeliverResult, error) {
+	msg, err := tx.GetMsg()
+	if err != nil {
+		return nil, err
+	}
+	perm := d.resolvePermission(msg)
+
+	if !d.authZ.HasPermission(ctx, perm) {
+		return nil, errors.Wrap(errors.ErrUnauthorized, "insufficient permissions")
+	}
+	return next.Deliver(ctx, store, tx)
+}
+
+func (d AuthZDecorator) resolvePermission(msg weave.Msg) Permission {
```

Comment:
Hmmm... I wonder about this...

The bonus is that the permissions are included on all modules without any change to their code (them checking for rbac permissions). But somehow this global filter based on Path is a bit odd. Maybe correct though.

What I understand here is that every message will have some auto-generated permission string. And no one can execute any message on the system without the corresponding role. This is secure, but is it usable?

What about creating a multisig? Maybe there are some msgs that should be universal?

Then again, if you have no fees, you want an anti-spam mechanism. So, something like making a "member" role that is a whitelist of all registered accounts, can have a few basic permissions (like create/update multisig) and then the more secure ones (like update grafain) belong to those special groups.

What do you think about unpermissioned messages or such a catch-all "members" whitelist?

Reply:
I had thought about adding wildcard permissions later. The message path is already a hierarchy so that anything under `_grafain.*` could be matched for example. There are definitively problems with this approach, when adding new messages or refactorings on the msg names/ path. 
I will give this some more thoughts. Good feedback 💐